### PR TITLE
`BasicRule_Generic` Circular Reasoning

### LINKED ISSUE #417: [BUG] Short Truth-Table: Atomic Rule Bug
### Description

The Atomic Rule checks out when you give the same truth-assignment to two or more instances of the same variable at the same time, i.e. in one step. It also happens with multiple variables, e.g. you set 2 A's and 2 B's all to True at the same time. Or when you set 3 A's all to False at the same time. And if you already have some A set to True, you can still set 2 or more A's to False ... the rule checks out as long as for each variable that you are giving a truth-value to, there are other instances you give that same truth-value to.

I believe that currently the rule checks out because the changes *self-justify each other*: If I set 2 A's to True, then the one A being set to True is justified by the other being set to True, and vice versa. We had a similar problem with the Must Light rule in LightUp (see Issue #119 )  ... is there something wrong with the general way the rule checking happens? Because the geenral rule is that *each* change should be checked individually, and as a change compared to the board as it is *before* the change

### Steps to Reproduce

Start any Short truth-table puzzle. As the first step, set two instances of the same variable to the same value. Now select Basic Rule-Atomic Rule. It will check out but it shouldn't.

### Expected Behavior

The rule should only check out when you *already* have an instance of that variable, and copy that value to another instance elsewhere. This is not what happens in described scenario": it should *not* check out.

### Screenshots

image

### Puzzle Files

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by 19690ao**
## Description

Unlike most puzzles' basic/direct rules, the rules for short truth tables are either direct or indirect subclasses of an abstract class called `BasicRule_Generic`. This rule happened to have a self-justification bug, which happened to affect every single basic rule in STT. 

Prof. Bram noticed the one that affected `BasicRuleAtomic` (#417), but that was one of many. 

For example, using biconditional elimination:
bi-el-2
...and a less obvious example using or elimination:
and-elim-circ
Although this transition is invalid, this bug is the reason that one of those cells is validated even though they're both equally part of the problem.

TLDR: Many STT basic rules have been bugged, possibly since its creation. I tried it on my build of Legup from LMS back from last year's intro to logic course, and I was able to reproduce the same behavior

oldversbugstt

Closes #417 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (improvement to an already existing feature)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by charlestian23 — CHANGES_REQUESTED** [review]
I just pushed a change to `dev` that refactors a lot of files. Can you pull that onto your branch and fix any merge conflicts?

**Review by charlestian23 — APPROVED** [review]
Changes seem good!
